### PR TITLE
Bind ExRouteAdapter instance to render callbacks

### PR DESCRIPTION
**Pull request by axelander**
https://github.com/aksonov/react-native-router-flux/issues/229

### CODE REVIEW

**Comment by aksonov** [comment]
Sorry, but it could brake existing apps. Why binding to this.route is not enough for rendering?

**Comment by axelander** [comment]
I'm quite positive this is a bug but I'm not sure where to fix it.
I found that in the constructor of `ExRouteAdapter`, [on this line](https://github.com/aksonov/react-native-router-flux/blob/master/ExRouter.js#L33), title is set correct for the `ExRouteAdapter` instance but **not** the `Route`. Updating my PR to show you.
I really have no idea how this should be solved, you think you could help me?

File: ExRouter.js [hunk 1]
```diff
@@ -40,7 +41,7 @@ export class ExRouteAdapter {
             this.renderRightButton = this.route.props.renderRightButton.bind(this.route);
         }
         if (this.route.props.renderTitle){
-            this.renderTitle = this.route.props.renderTitle.bind(this.route);
+            this.renderTitle = this.route.props.renderTitle.bind(this, this.route);
```
Comment (axelander) [inline ExRouter.js]: If route should be passed as the first argument to renderTitle then this update should be correct.

File: ExRouter.js [hunk 2]
```diff
@@ -30,6 +30,7 @@ export class ExRouteAdapter {
         }
         this.route = route;
         this.name = route.name;
+        if (props.title) this.route.title = props.title;
```
Comment (axelander) [inline ExRouter.js]: Feels like I'm correcting something that should be handled earlier on in some other module, but this solves the issue with getting the parent route title.

Comment (aksonov) [inline ExRouter.js]: Below we have 
```
this.title = props.title || route.title;
```

**Comment by aksonov** [comment]
Looks like the problem that Route doesn't update their props during actions. Then probably we have to fix BaseRouter. It should modify stack route with latest props, could you check there?

For example:
```
    _push(name:string, props:{ [key: string]: any} ){
        this._stack.push(name);
    }
```

Probably it should be added :
```
this.routes[name].props = Object.assign({}, this.routes[name].props, props);
```
What do you think? I don't have much time now to add such calls for all action types, check if everything is OK with demo, maybe someone could check.

**Comment by aksonov** [comment]
@danielweinmann Could you look it? I still don't like your fix when you rebuild whole route stack for new props coming from Router, maybe it would be more correct to fix BaseRouter and set new props after each action there?

File: ExRouter.js (hunk 2 above)
Comment (axelander) [inline ExRouter.js]: yes but this.title is on the `ExRouteAdapter` but the renderTitle callback is bound to `this.route`.

**Comment by danielweinmann** [comment]
@aksonov I'll try to take a look later this week!

About your suggestion, I don't think after new action is a good enough place to set new props. It's important that every react component re-renders when receiving new, different, props.

No matter what they know about their own state, thinks can change on the outside and shouldn't have to dispatch an action to change.

I'm still guessing the main reason you guys don't share my line of thinking is because you all must be using a big root router that manages a whole app. Is that right? In my case, I have a lot of state management without any route and a few small routers deep inside my rendering stack. This makes sense for me, since I only want to have a router when I need navigation functionality, and in a big part of my app I don't need navigation. It would mean a lot on unnecessary code to have a big root router.

So I need a way to do just that: rebuild the whole stack for new props coming from Router.

PS: I still don't like my solution in terms of elegance of the code. But it solves this scenario, and any other solution presented so far don't.

PS2: I think when @MITDelian finishes making BaseRouter immutable we will see more clearly how to make my solution better.

**Comment by aksonov** [comment]
Closing it for now because of 3.0 release is published.

**Pull request closed without merging**